File: typescript/src/llms.ts
/**
 * KServeLLM — LangChain LLM (text completion) for KServe inference services.
 *
 * Use this for base/completion models. For chat/instruct models, use ChatKServe.
 *
 * @example
 * ```typescript
 * import { KServeLLM } from "@langchain/kserve";
 *
 * const llm = new KServeLLM({
 *   baseUrl: "https://my-model.cluster.example.com",
 *   modelName: "my-base-model",
 *   protocol: "v2", // or "openai" for vLLM-served completion models
 * });
 *
 * const response = await llm.invoke("The capital of France is");
 * ```
 */

import { getEnvironmentVariable } from "@langchain/core/utils/env";
import { BaseLLM } from "@langchain/core/language_models/llms";
import type { CallbackManagerForLLMRun } from "@langchain/core/callbacks/manager";
import {
  GenerationChunk,
  LLMResult,
} from "@langchain/core/outputs";

import { KServeClient } from "./client.js";
import {
  buildCompletionRequest,
  parseCompletionResponse,
  parseCompletionStreamChunk,
} from "./openai-compat.js";
import {
  buildV2StreamRequest,
  buildV2TextRequest,
  getV2InferPath,
  parseV2TextResponse,
  parseV2TextStreamChunk,
} from "./v2-protocol.js";
import type {
  KServeLLMCallOptions,
  KServeLLMInput,
  KServeModelInfo,
  V2InferResponse,
} from "./types.js";
import { KServeInferenceError } from "./errors.js";

/**
 * KServeLLM — LangChain BaseLLM implementation for KServe text completion.
 */
export class KServeLLM extends BaseLLM<KServeLLMCallOptions> {
  static lc_name(): string {
    return "KServeLLM";
  }

  // Connection
  private readonly baseUrl: string;
  private readonly modelName: string;
  private readonly protocolPref: "openai" | "v2" | "auto";
  private readonly client: KServeClient;

  // Generation params
  private readonly temperature?: number;
  private readonly maxTokens?: number;
  private readonly topP?: number;
  private readonly stop?: string[];
  private readonly streaming: boolean;

  constructor(fields: KServeLLMInput) {
    super(fields);

    this.baseUrl =
      fields.baseUrl ??
      getEnvironmentVariable("KSERVE_BASE_URL") ??
      (() => {
        throw new Error(
          "baseUrl is required. Set it in the constructor or via KSERVE_BASE_URL env var."
        );
      })();

    this.modelName =
      fields.modelName ??
      getEnvironmentVariable("KSERVE_MODEL_NAME") ??
      (() => {
        throw new Error(
          "modelName is required. Set it in the constructor or via KSERVE_MODEL_NAME env var."
        );
      })();

    const protocolFromEnv = getEnvironmentVariable("KSERVE_PROTOCOL") as
      | "openai"
      | "v2"
      | "auto"
      | undefined;

    this.protocolPref = fields.protocol ?? protocolFromEnv ?? "auto";
    this.temperature = fields.temperature;
    this.maxTokens = fields.maxTokens;
    this.topP = fields.topP;
    this.stop = fields.stop;
    this.streaming = fields.streaming ?? false;

    this.client = new KServeClient({
      baseUrl: this.baseUrl,
      apiKey:
        fields.apiKey ?? getEnvironmentVariable("KSERVE_API_KEY"),
      tokenProvider: fields.tokenProvider,
      verifySsl: fields.verifySsl,
      caBundle: fields.caBundle ?? getEnvironmentVariable("KSERVE_CA_BUNDLE"),
      timeout: fields.timeout,
      maxRetries: fields.maxRetries,
    });

    if (this.protocolPref !== "auto") {
      this.client.setProtocol(this.protocolPref);
    }
  }

  /** LangChain model type identifier */
  _llmType(): string {
    return "kserve-llm";
  }

  // --------------------------------------------------------
  // Protocol resolution
  // --------------------------------------------------------

  private async resolveProtocol(): Promise<"openai" | "v2"> {
    if (this.protocolPref !== "auto") return this.protocolPref;
    return this.client.detectProtocol();
  }

  // --------------------------------------------------------
  // Core generation
  // --------------------------------------------------------

  /**
   * Generate completions for a batch of prompts.
   * Called by LangChain's `invoke()`, `generate()`, etc.
   */
  async _generate(
    prompts: string[],
    options: this["ParsedCallOptions"],
    runManager?: CallbackManagerForLLMRun
  ): Promise<LLMResult> {
    // For streaming, process prompts one at a time and collect chunks
    if (this.streaming) {
      const allGenerations: LLMResult["generations"] = [];
      let lastUsage: import("./types.js").OpenAIUsage | undefined = undefined;

      for (const prompt of prompts) {
        const chunks: GenerationChunk[] = [];
        for await (const chunk of this._streamResponseChunks(
          prompt,
          options,
          runManager
        )) {
          chunks.push(chunk);
        }
        const text = chunks.map((c) => c.text).join("");
        const lastGenInfo = chunks[chunks.length - 1]?.generationInfo as
          | (import("./types.js").KServeGenerationInfo & Record<string, unknown>)
          | undefined;
        if (lastGenInfo?.tokenUsage) {
          const tu = lastGenInfo.tokenUsage as {
            promptTokens: number;
            completionTokens: number;
            totalTokens: number;
          };
          lastUsage = {
            prompt_tokens: tu.promptTokens,
            completion_tokens: tu.completionTokens,
            total_tokens: tu.totalTokens,
          };
        }
        allGenerations.push([{ text, generationInfo: lastGenInfo }]);
      }

      return {
        generations: allGenerations,
        llmOutput: lastUsage
          ? {
              tokenUsage: {
                promptTokens: lastUsage.prompt_tokens,
                completionTokens: lastUsage.completion_tokens,
                totalTokens: lastUsage.total_tokens,
              },
            }
          : undefined,
      };
    }

    const protocol = await this.resolveProtocol();
    const allGenerations: LLMResult["generations"] = [];

    for (const prompt of prompts) {
      if (protocol === "openai") {
        const result = await this._generateOpenAI(prompt, options);
        allGenerations.push(...result.generations);
      } else {
        const result = await this._generateV2(prompt, options);
        allGenerations.push(...result.generations);
      }
    }

    return { generations: allGenerations };
  }

  private async _generateOpenAI(
    prompt: string,
    options: this["ParsedCallOptions"]
  ): Promise<LLMResult> {
    const request = buildCompletionRequest(
      this.modelName,
      prompt,
      {
        temperature: options.temperature ?? this.temperature,
        maxTokens: options.maxTokens ?? this.maxTokens,
        topP: this.topP,
        stop: options.stop ?? this.stop,
      },
      false
    );

    const response = await this.client.request<
      import("./types.js").OpenAICompletionResponse
    >("/v1/completions", request, { signal: options.signal });

    return parseCompletionResponse(response, this.modelName);
  }

  private async _generateV2(
    prompt: string,
    options: this["ParsedCallOptions"]
  ): Promise<LLMResult> {
    const request = buildV2TextRequest(prompt, {
      temperature: options.temperature ?? this.temperature,
      maxTokens: options.maxTokens ?? this.maxTokens,
      topP: this.topP,
      stop: options.stop ?? this.stop,
    });

    const response = await this.client.request<V2InferResponse>(
      getV2InferPath(this.modelName),
      request,
      { signal: options.signal }
    );

    return parseV2TextResponse(response, this.modelName);
  }

  // --------------------------------------------------------
  // Streaming
  // --------------------------------------------------------

  /**
   * Stream response chunks for a single prompt.
   */
  async *_streamResponseChunks(
    input: string,
    options: this["ParsedCallOptions"],
    runManager?: CallbackManagerForLLMRun
  ): AsyncGenerator<GenerationChunk> {
    const protocol = await this.resolveProtocol();

    if (protocol === "openai") {
      yield* this._streamOpenAI(input, options, runManager);
    } else {
      yield* this._streamV2(input, options, runManager);
    }
  }

  private async *_streamOpenAI(
    prompt: string,
    options: this["ParsedCallOptions"],
    runManager?: CallbackManagerForLLMRun
  ): AsyncGenerator<GenerationChunk> {
    const request = buildCompletionRequest(
      this.modelName,
      prompt,
      {
        temperature: options.temperature ?? this.temperature,
        maxTokens: options.maxTokens ?? this.maxTokens,
        topP: this.topP,
        stop: options.stop ?? this.stop,
      },
      true
    );

    for await (const raw of this.client.streamRequest(
      "/v1/completions",
      request,
      { signal: options.signal }
    )) {
      const parsed = parseCompletionStreamChunk(raw);
      if (parsed === null) continue;
      const { text, usage } = parsed;
      const genInfo: import("./types.js").KServeGenerationInfo & Record<string, unknown> = {
        protocol: "openai",
        modelName: this.modelName,
      };
      if (usage) {
        genInfo.tokenUsage = {
          promptTokens: usage.prompt_tokens,
          completionTokens: usage.completion_tokens,
          totalTokens: usage.total_tokens,
        };
      }
      if (text) {
        await runManager?.handleLLMNewToken(text);
      }
      yield new GenerationChunk({
        text,
        generationInfo: genInfo,
      });
    }
  }

  private async *_streamV2(
    prompt: string,
    options: this["ParsedCallOptions"],
    runManager?: CallbackManagerForLLMRun
  ): AsyncGenerator<GenerationChunk> {
    const baseRequest = buildV2TextRequest(prompt, {
      temperature: options.temperature ?? this.temperature,
      maxTokens: options.maxTokens ?? this.maxTokens,
      topP: this.topP,
      stop: options.stop ?? this.stop,
    });

    const streamRequest = buildV2StreamRequest(baseRequest);

    let streamed = false;
    try {
      for await (const raw of this.client.streamNDJSON(
        getV2InferPath(this.modelName),
        streamRequest,
        { signal: options.signal }
      )) {
        const chunk = parseV2TextStreamChunk(raw, this.modelName);
        if (!chunk) continue;
        streamed = true;
        await runManager?.handleLLMNewToken(chunk.text ?? "");
        yield chunk;
      }
    } catch (err) {
      if (!streamed) {
        console.warn(
          "[KServeLLM] V2 streaming not available, falling back to non-streaming"
        );
        const result = await this._generateV2(prompt, options);
        const gen = result.generations[0]?.[0];
        if (!gen) throw new KServeInferenceError("No generation in V2 response");
        await runManager?.handleLLMNewToken(gen.text);
        yield gen as GenerationChunk;
        return;
      }
      throw err;
    }
  }

  // --------------------------------------------------------
  // Model introspection
  // --------------------------------------------------------

  /**
   * Retrieve metadata about the model from the KServe endpoint.
   *
   * Tries the OpenAI-compatible /v1/models/{model} endpoint first,
   * then falls back to the V2 /v2/models/{model} endpoint.
   */
  async getModelInfo(): Promise<KServeModelInfo> {
    return this.client.getModelInfo(this.modelName);
  }
}

File: python/langchain_kserve/llms.py
"""KServeLLM — LangChain BaseLLM wrapper for KServe-hosted base completion models."""

from __future__ import annotations

import logging
from typing import (
    Any,
    AsyncIterator,
    Callable,
    Dict,
    Iterator,
    List,
    Literal,
    Optional,
    Tuple,
)

import httpx
from langchain_core.callbacks import (
    AsyncCallbackManagerForLLMRun,
    CallbackManagerForLLMRun,
)
from langchain_core.language_models import BaseLLM
from langchain_core.outputs import Generation, GenerationChunk, LLMResult
from langchain_core.utils import get_from_env
from pydantic import ConfigDict, Field, SecretStr, model_validator

from langchain_kserve._common import (
    KServeModelInfo,
    async_request_with_retry,
    build_async_client,
    build_sync_client,
    fetch_model_info_openai,
    fetch_model_info_v2,
    request_with_retry,
)
from langchain_kserve._openai_compat import (
    async_probe_openai_compat,
    astream_completion_response,
    build_completion_request,
    parse_completion_response,
    probe_openai_compat,
    stream_completion_response,
)
from langchain_kserve._v2_protocol import (
    build_v2_infer_request,
    infer_path,
    parse_v2_completion_response,
)

logger = logging.getLogger(__name__)


class KServeLLM(BaseLLM):
    """LangChain ``BaseLLM`` wrapper for KServe-hosted base/completion models.

    Use this class for raw (non-chat) models.  For instruction-tuned or chat
    models prefer :class:`~langchain_kserve.ChatKServe`.

    Supports both the OpenAI-compatible ``/v1/completions`` endpoint and the
    V2 Inference Protocol ``/v2/models/{model}/infer`` endpoint.

    Example:
        .. code-block:: python

            from langchain_kserve import KServeLLM

            llm = KServeLLM(
                base_url="https://llama.default.svc.cluster.local",
                model_name="llama-3-8b",
                max_tokens=256,
            )
            text = llm.invoke("Once upon a time")

    Environment variables:
        ``KSERVE_BASE_URL``, ``KSERVE_API_KEY``, ``KSERVE_MODEL_NAME``,
        ``KSERVE_PROTOCOL``, ``KSERVE_CA_BUNDLE``.
    """

    model_config = ConfigDict(arbitrary_types_allowed=True)

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------
    base_url: str = Field(default="", description="Root URL of the KServe service.")
    model_name: str = Field(default="", description="Model identifier.")
    protocol: Literal["openai", "v2", "auto"] = Field(
        default="auto",
        description="Inference protocol ('openai', 'v2', or 'auto').",
    )

    # ------------------------------------------------------------------
    # Auth
    # ------------------------------------------------------------------
    api_key: Optional[SecretStr] = Field(default=None)
    token_provider: Optional[Callable[[], str]] = Field(default=None, exclude=True)

    # ------------------------------------------------------------------
    # TLS
    # ------------------------------------------------------------------
    verify_ssl: bool = Field(default=True)
    ca_bundle: Optional[str] = Field(default=None)

    # ------------------------------------------------------------------
    # Generation params
    # ------------------------------------------------------------------
    temperature: float = Field(default=0.7, ge=0.0)
    max_tokens: Optional[int] = Field(default=None, ge=1)
    top_p: float = Field(default=1.0, ge=0.0, le=1.0)
    stop: Optional[List[str]] = Field(default=None)
    streaming: bool = Field(default=False)
    logprobs: Optional[bool] = Field(default=None)
    top_logprobs: Optional[int] = Field(default=None)

    # ------------------------------------------------------------------
    # Connection behaviour
    # ------------------------------------------------------------------
    timeout: int = Field(default=120, ge=1)
    max_retries: int = Field(default=3, ge=0)

    _resolved_protocol: Optional[Literal["openai", "v2"]] = None

    # ------------------------------------------------------------------
    # Validators
    # ------------------------------------------------------------------

    @model_validator(mode="before")
    @classmethod
    def _populate_from_env(cls, values: Dict[str, Any]) -> Dict[str, Any]:
        """Populate fields from environment variables if not explicitly set."""
        if not values.get("base_url"):
            values["base_url"] = get_from_env("base_url", "KSERVE_BASE_URL", default="")
        if not values.get("api_key"):
            env_key = get_from_env("api_key", "KSERVE_API_KEY", default="")
            if env_key:
                values["api_key"] = SecretStr(env_key)
        if not values.get("model_name"):
            values["model_name"] = get_from_env(
                "model_name", "KSERVE_MODEL_NAME", default=""
            )
        if values.get("protocol", "auto") == "auto":
            env_proto = get_from_env("protocol", "KSERVE_PROTOCOL", default="")
            if env_proto in ("openai", "v2"):
                values["protocol"] = env_proto
        if not values.get("ca_bundle"):
            values["ca_bundle"] = (
                get_from_env("ca_bundle", "KSERVE_CA_BUNDLE", default="") or None
            )
        return values

    # ------------------------------------------------------------------
    # Required LangChain property
    # ------------------------------------------------------------------

    @property
    def _llm_type(self) -> str:
        return "kserve-llm"

    # ------------------------------------------------------------------
    # HTTP client helpers
    # ------------------------------------------------------------------

    def _make_sync_client(self) -> httpx.Client:
        return build_sync_client(
            self.base_url,
            self.api_key,
            self.token_provider,
            self.verify_ssl,
            self.ca_bundle,
            self.timeout,
        )

    def _make_async_client(self) -> httpx.AsyncClient:
        return build_async_client(
            self.base_url,
            self.api_key,
            self.token_provider,
            self.verify_ssl,
            self.ca_bundle,
            self.timeout,
        )

    # ------------------------------------------------------------------
    # Protocol resolution
    # ------------------------------------------------------------------

    def _resolve_protocol_sync(self, client: httpx.Client) -> Literal["openai", "v2"]:
        if self.protocol != "auto":
            return self.protocol  # type: ignore[return-value]
        if self._resolved_protocol is not None:
            return self._resolved_protocol
        resolved: Literal["openai", "v2"] = (
            "openai" if probe_openai_compat(client, self.max_retries) else "v2"
        )
        object.__setattr__(self, "_resolved_protocol", resolved)
        return resolved

    async def _resolve_protocol_async(
        self, client: httpx.AsyncClient
    ) -> Literal["openai", "v2"]:
        if self.protocol != "auto":
            return self.protocol  # type: ignore[return-value]
        if self._resolved_protocol is not None:
            return self._resolved_protocol
        resolved: Literal["openai", "v2"] = (
            "openai"
            if await async_probe_openai_compat(client, self.max_retries)
            else "v2"
        )
        object.__setattr__(self, "_resolved_protocol", resolved)
        return resolved

    # ------------------------------------------------------------------
    # BaseLLM required: _generate
    # ------------------------------------------------------------------

    def _generate(
        self,
        prompts: List[str],
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> LLMResult:
        """Generate completions for a list of prompts synchronously.

        Args:
            prompts: List of text prompts.
            stop: Optional stop sequences.
            run_manager: Optional callback manager.
            **kwargs: Extra parameters.

        Returns:
            :class:`~langchain_core.outputs.LLMResult`.
        """
        generations: List[List[Generation]] = []
        last_usage: Optional[Dict[str, Any]] = None
        with self._make_sync_client() as client:
            proto = self._resolve_protocol_sync(client)
            effective_stop = stop or self.stop
            for prompt in prompts:
                text, usage = self._call_single(
                    client, proto, prompt, effective_stop, **kwargs
                )
                if usage is not None:
                    last_usage = usage
                generations.append([Generation(text=text)])
        llm_output: Optional[Dict[str, Any]] = (
            {"token_usage": last_usage} if last_usage is not None else None
        )
        return LLMResult(generations=generations, llm_output=llm_output)

    def _call_single(
        self,
        client: httpx.Client,
        proto: Literal["openai", "v2"],
        prompt: str,
        stop: Optional[List[str]],
        **kwargs: Any,
    ) -> Tuple[str, Optional[Dict[str, Any]]]:
        if proto == "openai":
            body = build_completion_request(
                model_name=self.model_name,
                prompt=prompt,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                top_p=self.top_p,
                stop=stop,
                stream=False,
                extra_kwargs=kwargs or None,
                logprobs=self.logprobs,
                top_logprobs=self.top_logprobs,
            )
            response = request_with_retry(
                client, "POST", "/v1/completions", self.max_retries, json=body
            )
            return parse_completion_response(response, self.model_name)
        else:
            body_v2 = build_v2_infer_request(
                model_name=self.model_name,
                prompt=prompt,
                temperature=self.temperature,
                max_tokens=self.max_tokens,
                top_p=self.top_p,
                stop=stop,
                extra_kwargs=kwargs or None,
            )
            response = request_with_retry(
                client,
                "POST",
                infer_path(self.model_name),
                self.max_retries,
                json=body_v2,
            )
            return parse_v2_completion_response(response), None

    # ------------------------------------------------------------------
    # Streaming — sync
    # ------------------------------------------------------------------

    def _stream(
        self,
        prompt: str,
        stop: Optional[List[str]] = None,
        run_manager: Optional[CallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> Iterator[GenerationChunk]:
        """Stream completions for a single prompt synchronously.

        Args:
            prompt: Text prompt.
            stop: Optional stop sequences.
            run_manager: Optional callback manager.
            **kwargs: Extra parameters.

        Yields:
            :class:`~langchain_core.outputs.GenerationChunk`.
        """
        with self._make_sync_client() as client:
            proto = self._resolve_protocol_sync(client)
            effective_stop = stop or self.stop

            if proto == "openai":
                body = build_completion_request(
                    model_name=self.model_name,
                    prompt=prompt,
                    temperature=self.temperature,
                    max_tokens=self.max_tokens,
                    top_p=self.top_p,
                    stop=effective_stop,
                    stream=True,
                    extra_kwargs=kwargs or None,
                    logprobs=self.logprobs,
                    top_logprobs=self.top_logprobs,
                )
                for text in stream_completion_response(
                    client, "/v1/completions", body, self.model_name
                ):
                    chunk = GenerationChunk(text=text)
                    if run_manager:
                        run_manager.on_llm_new_token(text, chunk=chunk)
                    yield chunk
            else:
                # V2 doesn't have native streaming; do a single request and yield one chunk
                text, _usage = self._call_single(client, "v2", prompt, effective_stop, **kwargs)
                chunk = GenerationChunk(text=text)
                if run_manager:
                    run_manager.on_llm_new_token(text, chunk=chunk)
                yield chunk

    # ------------------------------------------------------------------
    # Streaming — async
    # ------------------------------------------------------------------

    async def _astream(
        self,
        prompt: str,
        stop: Optional[List[str]] = None,
        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
        **kwargs: Any,
    ) -> AsyncIterator[GenerationChunk]:
        """Stream completions for a single prompt asynchronously.

        Args:
            prompt: Text prompt.
            stop: Optional stop sequences.
            run_manager: Optional callback manager.
            **kwargs: Extra parameters.

        Yields:
            :class:`~langchain_core.outputs.GenerationChunk`.
        """
        async with self._make_async_client() as client:
            proto = await self._resolve_protocol_async(client)
            effective_stop = stop or self.stop

            if proto == "openai":
                body = build_completion_request(
                    model_name=self.model_name,
                    prompt=prompt,
                    temperature=self.temperature,
                    max_tokens=self.max_tokens,
                    top_p=self.top_p,
                    stop=effective_stop,
                    stream=True,
                    extra_kwargs=kwargs or None,
                    logprobs=self.logprobs,
                    top_logprobs=self.top_logprobs,
                )
                async for text in astream_completion_response(
                    client, "/v1/completions", body, self.model_name
                ):
                    chunk = GenerationChunk(text=text)
                    if run_manager:
                        await run_manager.on_llm_new_token(text, chunk=chunk)
                    yield chunk
            else:
                response = await async_request_with_retry(
                    client,
                    "POST",
                    infer_path(self.model_name),
                    self.max_retries,
                    json=build_v2_infer_request(
                        model_name=self.model_name,
                        prompt=prompt,
                        temperature=self.temperature,
                        max_tokens=self.max_tokens,
                        top_p=self.top_p,
                        stop=effective_stop,
                        extra_kwargs=kwargs or None,
                    ),
                )
                text = parse_v2_completion_response(response)
                chunk = GenerationChunk(text=text)
                if run_manager:
                    await run_manager.on_llm_new_token(text, chunk=chunk)
                yield chunk

    # ------------------------------------------------------------------
    # Model introspection
    # ------------------------------------------------------------------

    async def get_model_info(self) -> KServeModelInfo:
        """Fetch metadata about the served model.

        Resolves the protocol (if set to ``"auto"``) and queries the appropriate
        endpoint for model metadata.

        Returns:
            :class:`~langchain_kserve._common.KServeModelInfo` with model metadata.

        Raises:
            KServeModelNotFoundError: If the model is not found.
            KServeInferenceError: On other errors.
        """
        async with self._make_async_client() as client:
            proto = await self._resolve_protocol_async(client)
            if proto == "openai":
                return await fetch_model_info_openai(
                    client, self.model_name, self.max_retries
                )
            else:
                return await fetch_model_info_v2(
                    client, self.model_name, self.max_retries
                )

    # ------------------------------------------------------------------
    # Identifying params
    # ------------------------------------------------------------------

    @property
    def _identifying_params(self) -> Dict[str, Any]:
        return {
            "model_name": self.model_name,
            "base_url": self.base_url,
            "protocol": self.protocol,
            "temperature": self.temperature,
            "max_tokens": self.max_tokens,
        }
